File: frameworks/rails/adsmedia.rb
# ADSMedia Ruby on Rails Integration
# Send emails via ADSMedia API from Rails applications
#
# gem 'httparty'

require 'httparty'
require 'json'

module ADSMedia
  API_BASE_URL = 'https://api.adsmedia.live/v1'.freeze

  class Client
    include HTTParty
    base_uri API_BASE_URL

    def initialize(api_key = nil, default_from_name: 'Rails')
      @api_key = api_key || ENV['ADSMEDIA_API_KEY'] || Rails.application.credentials.dig(:adsmedia, :api_key)
      @default_from_name = default_from_name

      raise ArgumentError, 'ADSMedia API key not configured' unless @api_key
    end

    # Send a single email
    def send_email(to:, subject:, html:, to_name: nil, from_name: nil, text: nil, reply_to: nil)
      payload = {
        to: to,
        subject: subject,
        html: html,
        from_name: from_name || @default_from_name
      }
      payload[:to_name] = to_name if to_name
      payload[:text] = text if text
      payload[:reply_to] = reply_to if reply_to

      request(:post, '/send', payload)
    end

    # Send batch emails
    def send_batch(recipients:, subject:, html:, text: nil, preheader: nil, from_name: nil)
      payload = {
        recipients: recipients,
        subject: subject,
        html: html,
        from_name: from_name || @default_from_name
      }
      payload[:text] = text if text
      payload[:preheader] = preheader if preheader

      request(:post, '/send/batch', payload)
    end

    # Check if email is suppressed
    def check_suppression(email)
      request(:get, '/suppressions/check', email: email)
    end

    # Test API connection
    def ping
      request(:get, '/ping')
    end

    # Get usage statistics
    def usage
      request(:get, '/account/usage')
    end

    private

    def request(method, endpoint, body = nil)
      options = {
        headers: {
          'Authorization' => "Bearer #{@api_key}",
          'Content-Type' => 'application/json',
          'Accept' => 'application/json'
        }
      }

      if method == :get && body
        options[:query] = body
      elsif body
        options[:body] = body.to_json
      end

      response = self.class.send(method, endpoint, options)
      result = JSON.parse(response.body, symbolize_names: true)

      unless result[:success]
        raise StandardError, result.dig(:error, :message) || 'ADSMedia API Error'
      end

      result[:data]
    end
  end

  # Rails configuration
  class << self
    attr_accessor :configuration

    def configure
      self.configuration ||= Configuration.new
      yield(configuration)
    end

    def client
      @client ||= Client.new(configuration&.api_key, default_from_name: configuration&.from_name || 'Rails')
    end
  end

  class Configuration
    attr_accessor :api_key, :from_name

    def initialize
      @api_key = ENV['ADSMEDIA_API_KEY']
      @from_name = 'Rails'
    end
  end
end

File: frameworks/nextjs/adsmedia.ts
/**
 * ADSMedia Next.js Integration
 * Server actions and API routes for Next.js
 */

const API_BASE_URL = 'https://api.adsmedia.live/v1';

interface SendEmailOptions {
  to: string;
  subject: string;
  html: string;
  toName?: string;
  fromName?: string;
  text?: string;
  replyTo?: string;
}

interface BatchRecipient {
  email: string;
  name?: string;
}

interface BatchEmailOptions {
  recipients: BatchRecipient[];
  subject: string;
  html: string;
  text?: string;
  preheader?: string;
  fromName?: string;
}

async function apiRequest(endpoint: string, method = 'GET', body?: any) {
  const apiKey = process.env.ADSMEDIA_API_KEY;
  
  if (!apiKey) {
    throw new Error('ADSMEDIA_API_KEY not configured');
  }

  const options: RequestInit = {
    method,
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    cache: 'no-store',
  };

  if (body) {
    options.body = JSON.stringify(body);
  }

  const response = await fetch(`${API_BASE_URL}${endpoint}`, options);
  const data = await response.json();

  if (!data.success) {
    throw new Error(data.error?.message || 'ADSMedia API Error');
  }

  return data.data;
}

// Server Actions (App Router)
export async function sendEmail(options: SendEmailOptions) {
  'use server';
  
  const payload: any = {
    to: options.to,
    subject: options.subject,
    html: options.html,
  };

  if (options.toName) payload.to_name = options.toName;
  if (options.fromName) payload.from_name = options.fromName;
  if (options.text) payload.text = options.text;
  if (options.replyTo) payload.reply_to = options.replyTo;

  return apiRequest('/send', 'POST', payload);
}

export async function sendBatchEmail(options: BatchEmailOptions) {
  'use server';
  
  const payload: any = {
    recipients: options.recipients,
    subject: options.subject,
    html: options.html,
  };

  if (options.fromName) payload.from_name = options.fromName;
  if (options.text) payload.text = options.text;
  if (options.preheader) payload.preheader = options.preheader;

  return apiRequest('/send/batch', 'POST', payload);
}

export async function checkSuppression(email: string) {
  'use server';
  return apiRequest(`/suppressions/check?email=${encodeURIComponent(email)}`);
}

export async function ping() {
  'use server';
  return apiRequest('/ping');
}

// Client for API Routes (Pages Router)
export function createADSMediaClient() {
  return {
    send: (options: SendEmailOptions) => {
      const payload: any = {
        to: options.to,
        subject: options.subject,
        html: options.html,
      };
      if (options.toName) payload.to_name = options.toName;
      if (options.fromName) payload.from_name = options.fromName;
      if (options.text) payload.text = options.text;
      if (options.replyTo) payload.reply_to = options.replyTo;
      
      return apiRequest('/send', 'POST', payload);
    },
    sendBatch: (options: BatchEmailOptions) => {
      const payload: any = {
        recipients: options.recipients,
        subject: options.subject,
        html: options.html,
      };
      if (options.fromName) payload.from_name = options.fromName;
      if (options.text) payload.text = options.text;
      if (options.preheader) payload.preheader = options.preheader;
      
      return apiRequest('/send/batch', 'POST', payload);
    },
    checkSuppression: (email: string) => 
      apiRequest(`/suppressions/check?email=${encodeURIComponent(email)}`),
    ping: () => apiRequest('/ping'),
    getUsage: () => apiRequest('/account/usage'),
  };
}

export default createADSMediaClient;

